presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java

```diff
@@ -0,0 +1,841 @@
+        this.determinismEvaluator = new com.facebook.presto.sql.relational.DeterminismEvaluator(metadata.getFunctionManager());
+
+        this.visitor = new Visitor();
+    }
+
+    public Type getType()
+    {
+        return expression.getType();
+    }
+
+    public Object evaluate()
+    {
+        checkState(!optimize, "evaluate() not allowed for optimizer");
+        return expression.accept(visitor, null);
+    }
+
+    @VisibleForTesting
+    public Object optimize(SymbolResolver inputs)
+    {
+        checkState(optimize, "evaluate(SymbolResolver) not allowed for interpreter");
+        return expression.accept(visitor, inputs);
+    }
+
+    private class Visitor
+            implements RowExpressionVisitor<Object, Object>
+    {
+        @Override
+        public Object visitInputReference(InputReferenceExpression node, Object context)
+        {
+            return node;
+        }
+
+        @Override
+        public Object visitConstant(ConstantExpression node, Object context)
+        {
+            // Unify output type if necessary
+            if (node.getType().getJavaType() == Slice.class && node.getValue() instanceof String) {
+                return utf8Slice((String) node.getValue());
+            }
+            return node.getValue();
+        }
+
+        @Override
+        public Object visitVariableReference(VariableReferenceExpression node, Object context)
+        {
+            if (context instanceof SymbolResolver) {
+                return ((SymbolResolver) context).getValue(new Symbol(node.getName()));
+            }
+            return node;
+        }
+
+        @Override
+        public Object visitCall(CallExpression node, Object context)
+        {
+            List<Type> argumentTypes = new ArrayList<>();
+            List<Object> argumentValues = new ArrayList<>();
+            for (RowExpression expression : node.getArguments()) {
+                Object value = expression.accept(this, context);
+                argumentValues.add(value);
+                argumentTypes.add(expression.getType());
+            }
+
+            FunctionHandle functionHandle = node.getFunctionHandle();
+
+            // Special casing for large constant array construction
+            if (functionHandle.getSignature().getName().toUpperCase().equals(ARRAY_CONSTRUCTOR)) {
+                SpecialCallResult result = tryHandleArrayConstructor(node, argumentValues);
+                if (result.isHandled()) {
+                    return result.getValue();
+                }
+            }
+
+            // Special casing for cast
+            if (isCast(functionHandle)) {
+                SpecialCallResult result = tryHandleCast(node, argumentValues);
+                if (result.isHandled()) {
+                    return result.getValue();
+                }
+            }
+
+            // Special casing for like
+            StandardFunctionResolution resolution = new StandardFunctionResolution(metadata.getFunctionManager());
```

Comment:
Maybe put `isCast` in `StandardFunctionResolution` as well and have this as a class variable.